### PR TITLE
De-coupling xdoc attention with bod hiding

### PR DESCRIPTION
**Pull request by Alvorecer721**

### CODE REVIEW

File: megatron/core/datasets/gpt_dataset.py [hunk 1]
```diff
@@ -743,16 +743,21 @@ def _get_ltor_masks_and_position_ids(
...
+                        # BOD can't attend to any token
+                        attention_mask[0, i, :] = 0
+                    
+                    if not cross_document_attention:
+                        # Skip first BOD as there's no previous document
+                        if i > 0:  
+                            # Block attention to previous document
+                            attention_mask[0, i:, :i] = 0 
+
+                            # Each token before BOD (except the first BOD) is the last token of a document.
```
Comment (martinjaggi) [inline megatron/core/datasets/gpt_dataset.py:759]: yes the comment makes sense. but relation to next code line confuses me

File: megatron/core/datasets/gpt_dataset.py [hunk 2]
```diff
@@ -743,16 +743,21 @@ def _get_ltor_masks_and_position_ids(
...
+                            loss_mask[i-1] = 0.0
```
Comment (martinjaggi) [inline megatron/core/datasets/gpt_dataset.py:760]: why mask the last token of each doc? did you mean to mask the `BoD` instead, at pos `i`? (or have EoD plus BoD in which case we can loss mask both?)

Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:760]: In the absence of an explicit EOD token, I simulate its behavior by masking the last token of each document. This ensures that the token immediately preceding the BOD token is always masked. Since BOD marks the start of a new document, the token before it effectively serves as the "EOD" token. As a result, whether an actual EOD token exists or not, it will be masked whenever cross_document_attention is disabled. You can check how upstream code implement EOD loss masking [here](https://github.com/swiss-ai/Megatron-LM/blob/d0e8aabc4b0e71400f12de70577a97a1b05f73ea/megatron/core/datasets/gpt_dataset.py#L727). 

The masking of the BOD token for loss computation is handled [here](https://github.com/swiss-ai/Megatron-LM/blob/d0e8aabc4b0e71400f12de70577a97a1b05f73ea/megatron/core/datasets/gpt_dataset.py#L248), in the same section where goldfish loss masking is applied.

Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:760]: If an EOD token is present, the `i-1` token corresponds to it and will be masked. Otherwise, if no EOD token exists, the last token of the previous document (still `i-1`) will be masked instead. This ensures that the model does not make predictions across document boundaries.

Comment (martinjaggi) [reply megatron/core/datasets/gpt_dataset.py:760]: let's be careful not to mask real tokens. no EoD in our pretrain. can make separate config flag outside of cross-doc

Comment (Alvorecer721) [reply megatron/core/datasets/gpt_dataset.py:760]: Sorry it took me some time to verify the correctness using vscode debugger to debug our Megatron codebase.

In my understanding, in Auto-regressive model, each token typically learns to predict the next token in the sequence. Since we are masking the loss for the last token by setting its loss mask to 0.0, you're telling the model: "Don't try to learn to predict what comes after this token."

Now, consider a simple example of sequence length = 5, and we set an breakpoint [here](https://github.com/swiss-ai/Megatron-LM/blob/d0e8aabc4b0e71400f12de70577a97a1b05f73ea/megatron/core/datasets/gpt_dataset.py#L206C13-L207C43), then we have
```python
tokens = [128000, 912, 3445, 198, 128000]
labels = [912, 3445, 198, 128000, 41546]
loss_mask = [1., 1., 1., 0., 1.]
```

Here's what happens at each position `if no cross attention mask`:

1. **Position 0:** Model sees 128000 → predicts 912 → loss applied (mask=1.0)
   * Model learns to predict 912 after seeing 128000

2. **Position 1:** Model sees 912 → predicts 3445 → loss applied (mask=1.0)
   * Model learns to predict 3445 after seeing 912

3. **Position 2:** Model sees 3445 → predicts 198 → loss applied (mask=1.0)
   * Model learns to predict 198 after seeing 3445

4. **Position 3:** Model sees 198 → predicts 128000 → NO loss applied (mask=0.0)
   * Model makes a prediction but doesn't learn from it
   * Important: 198 is still used as input, but the model doesn't update weights based on how well it predicted 128000

5. **Position 4:** Model sees 128000 → predicts 41546 → loss applied (mask=1.0)
   * Model learns to predict 41546 after seeing 128000

The key insight: The model still learns to predict the token "198" (when "198" appears as a label at position 2), but doesn't learn from the prediction it makes after seeing "198" as input.

Comment (martinjaggi) [reply megatron/core/datasets/gpt_dataset.py:760]: update, we now will have both `BoD` and `EoD` in the pretrain stream. need to adapt the code here - starting from upstream i guess, with a new config flag of cross-doc masking, and another new config flag for possibly hiding the BoD from attention (less crucial)
